ports-of-call/span.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef _PORTS_OF_CALL_SPAN_HPP_
+#define _PORTS_OF_CALL_SPAN_HPP_
+
+#include "ports-of-call/portability.hpp"
+
+#include <cassert>
+#include <cstddef>
+
+namespace PortsOfCall {
+
+// ================================================================================================
+// Heavily simplified relative to std::span, but provides a similar concept in a way
+// that's portable to GPUs.
+
+template <typename DataType>
+class span {
+ private:
+  DataType *ptr_{nullptr};
+  std::size_t size_{0};
+
+ public:
```

Comment:
As a corollary to the above, having some traits with verbose naming is also beneficial. As a suggestion:

```c++
using element_type = T;
using value_type = std::remove_cv_t<T>;
using index_type = std::ptrdiff_t; 
using difference_type = index_type;
using pointer = T*;
using reference = T&;
using iterator = T*;
using reverse_iterator =
    std::reverse_iterator<iterator>;
```

Reply:
These look like the standard typedefs for C++ containers, so that makes sense.

Reply:
I used the standard typedefs from `std::span` (https://en.cppreference.com/w/cpp/container/span), so it's slightly different from what you suggested.  I checked, and `std::reverse_iterator` is `constexpr` as of C++17, so that works.